scipy/sparse/csgraph/_laplacian.py

```diff
@@ -37,118 +41,510 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
-        csgraph = csgraph.astype(float)
+    if normed and (
+        np.issubdtype(csgraph.dtype, np.signedinteger)
+        or np.issubdtype(csgraph.dtype, np.uint)
+    ):
+        csgraph = csgraph.astype(np.float64)
+
+    if form == "array":
+        create_lap = (
+            _laplacian_sparse if isspmatrix(csgraph) else _laplacian_dense
+        )
+    else:
+        create_lap = (
+            _laplacian_sparse_flo
+            if isspmatrix(csgraph)
+            else _laplacian_dense_flo
+        )
 
-    create_lap = _laplacian_sparse if isspmatrix(csgraph) else _laplacian_dense
     degree_axis = 1 if use_out_degree else 0
-    lap, d = create_lap(csgraph, normed=normed, axis=degree_axis,
-                        copy=copy)
+
+    lap, d = create_lap(
+        csgraph,
+        normed=normed,
+        axis=degree_axis,
+        copy=copy,
+        form=form,
+        dtype=dtype,
+        symmetrized=symmetrized,
+    )
     if return_diag:
         return lap, d
     return lap
 
 
-def _setdiag_dense(A, d):
-    A.flat[::len(d)+1] = d
+def _setdiag_dense(m, d):
+    step = len(d) + 1
+    m.flat[::step] = d
+
+
+def _laplace(m, d):
+    return lambda v: v * d[:, np.newaxis] - m @ v
+
+
+def _laplace_normed(m, d, nd):
+    laplace = _laplace(m, d)
+    return (
+        lambda v: nd[:, np.newaxis] * laplace(v * nd[:, np.newaxis])
+           )
+
+
+def _laplace_sym(m, d):
+    return (
+        lambda v: v * d[:, np.newaxis]
+        - m @ v
+        - np.transpose(np.conjugate(np.transpose(np.conjugate(v)) @ m))
+    )
+
+
+def _laplace_normed_sym(m, d, nd):
+    laplace_sym = _laplace_sym(m, d)
+    return (
+        lambda v: nd[:, np.newaxis] * laplace_sym(v * nd[:, np.newaxis])
+           )
+
+
+def _linearoperator(mv, shape, dtype):
+    return LinearOperator(matvec=mv, matmat=mv, shape=shape, dtype=dtype)
+
+
+def _laplacian_sparse_flo(graph, normed, axis, copy, form, dtype, symmetrized):
+    # The keyword argument ``copy`` is unused and has no effect here.
+    del copy  # Unused argument
+
+    if dtype is None:
+        dtype = graph.dtype
+
+    graph_sum = graph.sum(axis=axis).getA1()
+    graph_diagonal = graph.diagonal()
+    diag = graph_sum - graph_diagonal
+    if symmetrized:
+        graph_sum += graph.sum(axis=1 - axis).getA1()
+        diag = graph_sum - graph_diagonal - graph_diagonal
+
+    if normed:
+        isolated_node_mask = diag == 0
+        w = np.where(isolated_node_mask, 1, np.sqrt(diag))
+        if symmetrized:
+            md = _laplace_normed_sym(graph, graph_sum, 1.0 / w)
+        else:
+            md = _laplace_normed(graph, graph_sum, 1.0 / w)
+        if form == "function":
+            return md, w.astype(dtype, copy=False)
+        elif form == "lo":
+            m = _linearoperator(md, shape=graph.shape, dtype=dtype)
+            return m, w.astype(dtype, copy=False)
+        else:
+            raise ValueError(f"Invalid form: {repr(form)}")
```

Comment:
The new validation errors need to be tested using `pytest.raises(ValueError, match=...)`

Reply:
@tupui Could you please help and propose the specific change here? I'll then follow your sample in other occurrences.

Reply:
What I mean is that in the tests you need to call the function in a way that would make this warning getting raised. In the tests you would have:
```python
with pytest.raises(ValueError, match="Invalid form: 'toto'"):
    laplacian(..., form='toto')
```

By the way, noticing this now. Instead of `f"{repr(form)}"` what you want to do is `f"{form!r}"`.

Reply:
But the line you comment on is not in the test?

Reply:
It's a new message no? If I am not mistaking I did not see any test using a wrong `form` and checking the correct message is raised.

Reply:
got it, will do, thanks!

Reply:
Thanks 😃